### ISSUE TITLE
Bug: [SYCL] Backend seems not stable

### ISSUE THREAD

**Issue by qnixsynapse** [ggerganov/llama.cpp#8569]
### What happened?

I'm experiencing an issue with non Q4_K quants, where the output of my model gets broken or cut off before completion. This seems to happen especially when the model context is filled up with a lot of long messages. I ran `./test-backend-ops -b SYCL0` to test the backend, and found that it is failing on some operations such as SILU as shown in the log.

### Name and Version

$ ./llama-cli --version
version: 3414 (0d2c7321)
built with Intel(R) oneAPI DPC++/C++ Compiler 2024.1.0 (2024.1.0.20240308) for x86_64-unknown-linux-gnu

### What operating system are you seeing the problem on?

Linux

### Relevant log output

```shell
MUL_MAT(type_a=iq4_nl,type_b=f32,m=16,n=1,k=256,bs=[1,1],nr=[1,1]): [MUL_MAT] NMSE = 0.274506591 > 0.000500000 FAIL
MUL_MAT_ID(type_a=iq4_nl,type_b=f32,n_mats=4,n_used=2,b=0,m=512,n=1,k=256): [MUL_MAT_ID] NMSE = 0.291578637 > 0.000500000 FAIL
SILU(type=f32,ne_a=[7,13,19,23],v=0): [SILU] inf mismatch: SYCL0=-0.000000 CPU=-nan FAIL

...
...
1331/1334 tests passed
Backend SYCL0: FAIL
```

**Comment by qnixsynapse** [comment]
I'm wondering why the SILU operation test is failing. This operation is used in LLaMA3, and its failure might explain the unstable generation midway through the context that I'm experiencing.

```
$ ./test-backend-ops test -o SILU -b SYCL0 
ggml_sycl_init: GGML_SYCL_FORCE_MMQ:   no
ggml_sycl_init: SYCL_USE_XMX: yes
ggml_sycl_init: found 1 SYCL devices:
Testing 2 backends

Backend 1/2 (CPU)
  Skipping
Backend 2/2 (SYCL0)
[SYCL] call ggml_check_sycl
ggml_check_sycl: GGML_SYCL_DEBUG: 0
ggml_check_sycl: GGML_SYCL_F16: yes
found 1 SYCL devices:
|  |                   |                                       |       |Max    |        |Max  |Global |                     |
|  |                   |                                       |       |compute|Max work|sub  |mem    |                     |
|ID|        Device Type|                                   Name|Version|units  |group   |group|size   |       Driver version|
|--|-------------------|---------------------------------------|-------|-------|--------|-----|-------|---------------------|
| 0| [level_zero:gpu:0]|                Intel Arc A750 Graphics|    1.3|    448|    1024|   32|  8096M|            1.3.29735|
  Backend name: SYCL0
  SILU(type=f32,ne_a=[128,10,10,10],v=0): OK
  SILU(type=f32,ne_a=[7,13,19,23],v=0): [SILU] inf mismatch: SYCL0=-0.000000 CPU=-nan FAIL
  SILU(type=f32,ne_a=[128,10,10,10],v=1): not supported [SYCL0] 
  SILU(type=f32,ne_a=[7,13,19,23],v=1): not supported [SYCL0] 
  1333/1334 tests passed
  Backend SYCL0: FAIL

1/2 backends passed
FAIL
```

**Comment by slaren** [comment]
Since it is the CPU backend generating a `-nan` here, the issue is likely in the CPU backend. My guess is that recent changes to the silu implementation do not work with the Intel C++ compiler.

**Comment by qnixsynapse** [comment]
Update: I found out that models generally have a very high perplexity when kv cache is also offloaded to GPU(SYCL0). When KV cache is not offloaded, the models generation quality improves. I am not sure why this is happening.

**Comment by qnixsynapse** [comment]
Reverting the warp size to 32 fixes this issue at the cost of slight performance loss (seems unrelated to the SILU op failing in backend test and I don't use iq4_nl)

The backend seems really stable now even after crossing 1000 tokens in KV-cache.
Is this because warp size of 16 is too low for large matrices?

cc: @luoyu-intel @airMeng

**Comment by luoyu-intel** [comment]
Do you mean that if the context length is larger than 1000 the model will produce the wrong tokens?

**Comment by qnixsynapse** [comment]
Yes. That's what I am seeing here in my tests.

**Comment by luoyu-intel** [comment]
I see. Some kernels have two implementations depending on whether the parallel size is less than MAX_GROUP_SIZE(1024) or not. The larger case is not covered in the UTs. I will locate the bug kernel and add the UT case.

**Comment by qnixsynapse** [comment]
@luoyu-intel Thank you!

**Comment by luoyu-intel** [comment]
@qnixsynapse Recently there will be an optimization PR for Q4_0, >30% speedup.  There may be more in the future. When the Q4_0 is fully optimized, I will start considering the support of Q_K types.
For this issue, I locate that it's caused by the softmax kernel. I may need one or two days to fix it.

**Comment by qnixsynapse** [comment]
Please take your time.


Interesting that softmax op passes in test-backend-ops which only test for GGML_TYPE_F32 and not GGML_TYPE_F16 for some reason. Are the logits in F32 even when F16 is enabled? I did not test for F32 because in my card, F16 is the fastest according to the tests.

https://github.com/ggerganov/llama.cpp/blob/6f11a83e4e7700fdf353ed4a29599cb662c792f6/tests/test-backend-ops.cpp#L2344-L2345


https://github.com/ggerganov/llama.cpp/blob/6f11a83e4e7700fdf353ed4a29599cb662c792f6/tests/test-backend-ops.cpp#L2351-L2354


```
$ ./test-backend-ops -b SYCL0 -o SOFT_MAX
ggml_backend_register: registered backend CPU
ggml_sycl_init: GGML_SYCL_FORCE_MMQ:   no
ggml_sycl_init: SYCL_USE_XMX: yes
ggml_sycl_init: found 1 SYCL devices:
ggml_backend_register: registered backend SYCL0
Testing 2 backends

Backend 1/2 (CPU)
  Skipping
Backend 2/2 (SYCL0)
[SYCL] call ggml_check_sycl
ggml_check_sycl: GGML_SYCL_DEBUG: 0
ggml_check_sycl: GGML_SYCL_F16: yes
found 1 SYCL devices:
|  |                   |                                       |       |Max    |        |Max  |Global |                     |
|  |                   |                                       |       |compute|Max work|sub  |mem    |                     |
|ID|        Device Type|                                   Name|Version|units  |group   |group|size   |       Driver version|
|--|-------------------|---------------------------------------|-------|-------|--------|-----|-------|---------------------|
| 0| [level_zero:gpu:0]|                Intel Arc A750 Graphics|    1.3|    448|    1024|   32|  8096M|            1.3.29735|
  Backend name: SYCL0
  SOFT_MAX(type=f32,ne=[16,16,1,1],mask=0,scale=1.000000,max_bias=0.000000): OK
  SOFT_MAX(type=f32,ne=[15,15,1,1],mask=0,scale=1.000000,max_bias=0.000000): OK
  SOFT_MAX(type=f32,ne=[16,1024,1,1],mask=0,scale=1.000000,max_bias=0.000000): OK
  SOFT_MAX(type=f32,ne=[15,1023,1,1],mask=0,scale=1.000000,max_bias=0.000000): OK
  SOFT_MAX(type=f32,ne=[1024,16,1,1],mask=0,scale=1.000000,max_bias=0.000000): OK
  SOFT_MAX(type=f32,ne=[1023,15,1,1],mask=0,scale=1.000000,max_bias=0.000000): OK
  SOFT_MAX(type=f32,ne=[1024,1024,1,1],mask=0,scale=1.000000,max_bias=0.000000): OK
  SOFT_MAX(type=f32,ne=[1023,1023,1,1],mask=0,scale=1.000000,max_bias=0.000000): OK
  SOFT_MAX(type=f32,ne=[16,16,1,1],mask=0,scale=0.100000,max_bias=0.000000): OK
  SOFT_MAX(type=f32,ne=[15,15,1,1],mask=0,scale=0.100000,max_bias=0.000000): OK
  SOFT_MAX(type=f32,ne=[16,1024,1,1],mask=0,scale=0.100000,max_bias=0.000000): OK
  SOFT_MAX(type=f32,ne=[15,1023,1,1],mask=0,scale=0.100000,max_bias=0.000000): OK
  SOFT_MAX(type=f32,ne=[1024,16,1,1],mask=0,scale=0.100000,max_bias=0.000000): OK
  SOFT_MAX(type=f32,ne=[1023,15,1,1],mask=0,scale=0.100000,max_bias=0.000000): OK
  SOFT_MAX(type=f32,ne=[1024,1024,1,1],mask=0,scale=0.100000,max_bias=0.000000): OK
  SOFT_MAX(type=f32,ne=[1023,1023,1,1],mask=0,scale=0.100000,max_bias=0.000000): OK
  SOFT_MAX(type=f32,ne=[16,16,1,1],mask=1,scale=1.000000,max_bias=0.000000): OK
  SOFT_MAX(type=f32,ne=[15,15,1,1],mask=1,scale=1.000000,max_bias=0.000000): OK
  SOFT_MAX(type=f32,ne=[16,1024,1,1],mask=1,scale=1.000000,max_bias=0.000000): OK
  SOFT_MAX(type=f32,ne=[15,1023,1,1],mask=1,scale=1.000000,max_bias=0.000000): OK
  SOFT_MAX(type=f32,ne=[1024,16,1,1],mask=1,scale=1.000000,max_bias=0.000000): OK
  SOFT_MAX(type=f32,ne=[1023,15,1,1],mask=1,scale=1.000000,max_bias=0.000000): OK
  SOFT_MAX(type=f32,ne=[1024,1024,1,1],mask=1,scale=1.000000,max_bias=0.000000): OK
  SOFT_MAX(type=f32,ne=[1023,1023,1,1],mask=1,scale=1.000000,max_bias=0.000000): OK
  SOFT_MAX(type=f32,ne=[16,16,1,1],mask=1,scale=0.100000,max_bias=0.000000): OK
  SOFT_MAX(type=f32,ne=[15,15,1,1],mask=1,scale=0.100000,max_bias=0.000000): OK
  SOFT_MAX(type=f32,ne=[16,1024,1,1],mask=1,scale=0.100000,max_bias=0.000000): OK
  SOFT_MAX(type=f32,ne=[15,1023,1,1],mask=1,scale=0.100000,max_bias=0.000000): OK
  SOFT_MAX(type=f32,ne=[1024,16,1,1],mask=1,scale=0.100000,max_bias=0.000000): OK
  SOFT_MAX(type=f32,ne=[1023,15,1,1],mask=1,scale=0.100000,max_bias=0.000000): OK
  SOFT_MAX(type=f32,ne=[1024,1024,1,1],mask=1,scale=0.100000,max_bias=0.000000): OK
  SOFT_MAX(type=f32,ne=[1023,1023,1,1],mask=1,scale=0.100000,max_bias=0.000000): OK
  SOFT_MAX(type=f32,ne=[16,16,1,1],mask=1,scale=1.000000,max_bias=8.000000): OK
  SOFT_MAX(type=f32,ne=[15,15,1,1],mask=1,scale=1.000000,max_bias=8.000000): OK
  SOFT_MAX(type=f32,ne=[16,1024,1,1],mask=1,scale=1.000000,max_bias=8.000000): OK
  SOFT_MAX(type=f32,ne=[15,1023,1,1],mask=1,scale=1.000000,max_bias=8.000000): OK
  SOFT_MAX(type=f32,ne=[1024,16,1,1],mask=1,scale=1.000000,max_bias=8.000000): OK
  SOFT_MAX(type=f32,ne=[1023,15,1,1],mask=1,scale=1.000000,max_bias=8.000000): OK
  SOFT_MAX(type=f32,ne=[1024,1024,1,1],mask=1,scale=1.000000,max_bias=8.000000): OK
  SOFT_MAX(type=f32,ne=[1023,1023,1,1],mask=1,scale=1.000000,max_bias=8.000000): OK
  SOFT_MAX(type=f32,ne=[16,16,1,1],mask=1,scale=0.100000,max_bias=8.000000): OK
  SOFT_MAX(type=f32,ne=[15,15,1,1],mask=1,scale=0.100000,max_bias=8.000000): OK
  SOFT_MAX(type=f32,ne=[16,1024,1,1],mask=1,scale=0.100000,max_bias=8.000000): OK
  SOFT_MAX(type=f32,ne=[15,1023,1,1],mask=1,scale=0.100000,max_bias=8.000000): OK
  SOFT_MAX(type=f32,ne=[1024,16,1,1],mask=1,scale=0.100000,max_bias=8.000000): OK
  SOFT_MAX(type=f32,ne=[1023,15,1,1],mask=1,scale=0.100000,max_bias=8.000000): OK
  SOFT_MAX(type=f32,ne=[1024,1024,1,1],mask=1,scale=0.100000,max_bias=8.000000): OK
  SOFT_MAX(type=f32,ne=[1023,1023,1,1],mask=1,scale=0.100000,max_bias=8.000000): OK
  SOFT_MAX(type=f32,ne=[16,2,32,1],mask=1,scale=0.100000,max_bias=0.000000): OK
  SOFT_MAX(type=f32,ne=[16,2,32,1],mask=0,scale=0.100000,max_bias=0.000000): OK
  SOFT_MAX(type=f32,ne=[32,2,32,1],mask=1,scale=0.100000,max_bias=0.000000): OK
  SOFT_MAX(type=f32,ne=[32,2,32,1],mask=1,scale=0.100000,max_bias=8.000000): OK
  1334/1334 tests passed
  Backend SYCL0: OK
```

**Comment by qnixsynapse** [comment]
Fixed in #8642 closing!
